File: chapter_4_recursionanddp/Problem_15_JumpGame.java
package chapter_4_recursionanddp;

public class Problem_15_JumpGame {

	public static int jump(int[] arr) {
		if (arr == null || arr.length == 0) {
			return 0;
		}
		int jump = 0;
		int cur = 0;
		int next = 0;
		for (int i = 0; i < arr.length; i++) {
			if (cur < i) {
				jump++;
				cur = next;
			}
			next = Math.max(next, i + arr[i]);
		}
		return jump;
	}

	public static void main(String[] args) {
		int[] arr = { 3, 2, 3, 1, 1, 4 };
		System.out.println(jump(arr));

	}

}

File: chapter_4_recursionanddp/Problem_15_JumpGame.cpp
#include <iostream>
#include <vector>

using namespace std;

int jump(vector<int> arr)
{
    if(arr.empty())
        return 0;

    int jump = 0;
    int cur = 0;
    int next = 0;
    for(int i = 0; i < arr.size(); i++)
    {
        if(cur < i)
        {
            jump++;
            cur = next;
        }
        next = max(next, i + arr[i]);
    }
    return jump;
}

int main()
{
    vector<int> arr = { 3, 2, 3, 1, 1, 4 };
    cout << jump(arr) << endl;
    return 0;
}
